src/librustc_mir/interpret/terminator.rs

```diff
@@ -388,12 +388,10 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpretCx<'mir, 'tcx, M> {
                             ));
                         }
                     } else {
-                        let callee_layout =
-                            self.layout_of_local(self.frame(), mir::RETURN_PLACE, None)?;
-                        if !callee_layout.abi.is_uninhabited() {
-                            return err!(FunctionRetMismatch(
-                                self.tcx.types.never, callee_layout.ty
-                            ));
+                        let local = mir::RETURN_PLACE;
+                        let ty = self.frame().body.local_decls[local].ty;
+                        if !self.tcx.is_ty_uninhabited_from_any_module(ty) {
```

Comment:
It seems rather strange to me that MIR semantics should depend on module boundaries or visibility or whatever this does... why is this needed? A comment would have been nice.

Reply:
@varkor this has landed now, could you follow-up with a comment that explains this?

Generally the layout of a type should contain all the information needed foe executing a program. We should *never* look at types. (Validity checks being an exception.)